In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn import svm
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report


from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.model_selection import train_test_split, cross_val_score



In [2]:
fire=pd.read_csv('forestfires.csv')
fire

month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain  ...  monthfeb  \
0     mar  fri  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0  ...         0   
1     oct  tue  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0  ...         0   
2     oct  sat  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0  ...         0   
3     mar  fri  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2  ...         0   
4     mar  sun  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0  ...         0   
..    ...  ...   ...    ...    ...   ...   ...  ..   ...   ...  ...       ...   
512   aug  sun  81.6   56.7  665.6   1.9  27.8  32   2.7   0.0  ...         0   
513   aug  sun  81.6   56.7  665.6   1.9  21.9  71   5.8   0.0  ...         0   
514   aug  sun  81.6   56.7  665.6   1.9  21.2  70   6.7   0.0  ...         0   
515   aug  sat  94.4  146.0  614.7  11.3  25.6  42   4.0   0.0  ...         0   
516   nov  tue  79.5    3.0  106.7   1.1  11.8  31   4.5   0.0  ...         0   

     monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  monthoct  \
0           0         0         0         1         0         0         0   
1           0         0         0         0         0         0         1   
2           0         0         0         0         0         0         1   
3           0         0         0         1         0         0         0   
4           0         0         0         1         0         0         0   
..        ...       ...       ...       ...       ...       ...       ...   
512         0         0         0         0         0         0         0   
513         0         0         0         0         0         0         0   
514         0         0         0         0         0         0         0   
515         0         0         0         0         0         0         0   
516         0         0         0         0         0         1         0   

     monthsep  size_category  
0           0          small  
1           0          small  
2           0          small  
3           0          small  
4           0          small  
..        ...            ...  
512         0          large  
513         0          large  
514         0          large  
515         0          small  
516         0          small  

[517 rows x 31 columns]

In [3]:
fire.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517 entries, 0 to 516
Data columns (total 31 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   month          517 non-null    object 
 1   day            517 non-null    object 
 2   FFMC           517 non-null    float64
 3   DMC            517 non-null    float64
 4   DC             517 non-null    float64
 5   ISI            517 non-null    float64
 6   temp           517 non-null    float64
 7   RH             517 non-null    int64  
 8   wind           517 non-null    float64
 9   rain           517 non-null    float64
 10  area           517 non-null    float64
 11  dayfri         517 non-null    int64  
 12  daymon         517 non-null    int64  
 13  daysat         517 non-null    int64  
 14  daysun         517 non-null    int64  
 15  daythu         517 non-null    int64  
 16  daytue         517 non-null    int64  
 17  daywed         517 non-null    int64  
 18  monthapr  

In [4]:
fire.isna().sum()

month            0
day              0
FFMC             0
DMC              0
DC               0
ISI              0
temp             0
RH               0
wind             0
rain             0
area             0
dayfri           0
daymon           0
daysat           0
daysun           0
daythu           0
daytue           0
daywed           0
monthapr         0
monthaug         0
monthdec         0
monthfeb         0
monthjan         0
monthjul         0
monthjun         0
monthmar         0
monthmay         0
monthnov         0
monthoct         0
monthsep         0
size_category    0
dtype: int64

In [6]:
fire=fire.drop(['monthjan','monthfeb','monthmar','monthapr','monthmay','monthjun','monthjul','monthaug','monthsep',
                'monthoct','monthnov','monthdec'], axis=1)

In [8]:
fire=fire.drop(['dayfri','daymon','daysat','daysun','daythu','daytue','daywed'], axis=1)
fire

month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain   area  \
0     mar  fri  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0   0.00   
1     oct  tue  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0   0.00   
2     oct  sat  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0   0.00   
3     mar  fri  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2   0.00   
4     mar  sun  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0   0.00   
..    ...  ...   ...    ...    ...   ...   ...  ..   ...   ...    ...   
512   aug  sun  81.6   56.7  665.6   1.9  27.8  32   2.7   0.0   6.44   
513   aug  sun  81.6   56.7  665.6   1.9  21.9  71   5.8   0.0  54.29   
514   aug  sun  81.6   56.7  665.6   1.9  21.2  70   6.7   0.0  11.16   
515   aug  sat  94.4  146.0  614.7  11.3  25.6  42   4.0   0.0   0.00   
516   nov  tue  79.5    3.0  106.7   1.1  11.8  31   4.5   0.0   0.00   

    size_category  
0           small  
1           small  
2           small  
3           small  
4           small  
..            ...  
512         large  
513         large  
514         large  
515         small  
516         small  

[517 rows x 12 columns]

In [11]:
from sklearn.preprocessing import LabelEncoder

In [12]:
col=['month','day','size_category']
fire[col]=fire[col].apply(LabelEncoder().fit_transform)

In [13]:
fire

month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain   area  \
0        7    0  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0   0.00   
1       10    5  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0   0.00   
2       10    2  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0   0.00   
3        7    0  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2   0.00   
4        7    3  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0   0.00   
..     ...  ...   ...    ...    ...   ...   ...  ..   ...   ...    ...   
512      1    3  81.6   56.7  665.6   1.9  27.8  32   2.7   0.0   6.44   
513      1    3  81.6   56.7  665.6   1.9  21.9  71   5.8   0.0  54.29   
514      1    3  81.6   56.7  665.6   1.9  21.2  70   6.7   0.0  11.16   
515      1    2  94.4  146.0  614.7  11.3  25.6  42   4.0   0.0   0.00   
516      9    5  79.5    3.0  106.7   1.1  11.8  31   4.5   0.0   0.00   

     size_category  
0                1  
1                1  
2                1  
3                1  
4                1  
..             ...  
512              0  
513              0  
514              0  
515              1  
516              1  

[517 rows x 12 columns]

In [14]:
x=fire.iloc[:, :-1]
y=fire.iloc[:, -1]

In [15]:
x

month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain   area
0        7    0  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0   0.00
1       10    5  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0   0.00
2       10    2  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0   0.00
3        7    0  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2   0.00
4        7    3  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0   0.00
..     ...  ...   ...    ...    ...   ...   ...  ..   ...   ...    ...
512      1    3  81.6   56.7  665.6   1.9  27.8  32   2.7   0.0   6.44
513      1    3  81.6   56.7  665.6   1.9  21.9  71   5.8   0.0  54.29
514      1    3  81.6   56.7  665.6   1.9  21.2  70   6.7   0.0  11.16
515      1    2  94.4  146.0  614.7  11.3  25.6  42   4.0   0.0   0.00
516      9    5  79.5    3.0  106.7   1.1  11.8  31   4.5   0.0   0.00

[517 rows x 11 columns]

In [16]:
y

0      1
1      1
2      1
3      1
4      1
      ..
512    0
513    0
514    0
515    1
516    1
Name: size_category, Length: 517, dtype: int32

In [18]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.25, random_state=42)

#### Grid Search CV

In [19]:
clf=SVC()
param_grid=[{'kernel':['rbf'], 'gamma':[50,5,10,0.5],'C':[15,14,13,12,11,10,0.1,0.001]}]
gsv=GridSearchCV(clf,param_grid,cv=10)
gsv.fit(x_train,y_train)

GridSearchCV(cv=10, estimator=SVC(),
             param_grid=[{'C': [15, 14, 13, 12, 11, 10, 0.1, 0.001],
                          'gamma': [50, 5, 10, 0.5], 'kernel': ['rbf']}])

In [20]:
gsv.best_params_,gsv.best_score_

({'C': 15, 'gamma': 50, 'kernel': 'rbf'}, 0.7365047233468286)

In [22]:
clf=SVC(C=15,gamma=50)
clf.fit(x_train,y_train)
y_pred=clf.predict(x_test)
acc=accuracy_score(y_test,y_pred)*100
print('Accuracy=',acc)
confusion_matrix(y_test,y_pred)


Accuracy= 73.07692307692307


array([[ 0, 35],
       [ 0, 95]], dtype=int64)

In [23]:
clf=SVC(kernel='linear')
clf.fit(x_train,y_train)
y_pred=clf.predict(x_test)
acc=accuracy_score(y_test,y_pred)*100
print('Accuracy=',acc)
confusion_matrix(y_test,y_pred)

Accuracy= 97.6923076923077


array([[33,  2],
       [ 1, 94]], dtype=int64)

##### when we apply linear kernel transformation we get accuracy 97.69%

In [24]:
clf = SVC(kernel= "poly") 
clf.fit(x_train , y_train)
y_pred = clf.predict(x_test)
acc = accuracy_score(y_test, y_pred) * 100
print("Accuracy =", acc)
confusion_matrix(y_test, y_pred) 

Accuracy = 78.46153846153847


array([[ 7, 28],
       [ 0, 95]], dtype=int64)

##### when we apply polynomial kernel transformation we get accuracy 78.46%